Fixes for Issues #43 and #44

jfdenise opened the pull request:
Constant string transformation (#44) and on fix for Annotation Element value of type String (#43).

A user would define a master index such as:
org/jboss/as/jaxrs/JaxrsAnnotations$Constants.class=jaxrs-annotations.properties
org/apache/jasper/compiler/Generator.class=jsp-compiler.properties
org/jboss/as/weld/CdiAnnotations$Constants.class=cdi-annotations.properties

Each properties file would then contain specific mapping for the class.
Note that the transformation also looks inside complex String to replace parts.

Added unit test.

Signed-off-by: JF Denise <jdenise@redhat.com>

tbitonti commented on org.eclipse.transformer/src/main/java/org/eclipse/transformer/Transformer.java:
```diff
@@ -945,6 +950,33 @@ public boolean setRules() throws IOException, URISyntaxException, IllegalArgumen
+                        }
+
+                        perClassConstantStrings = masterUpdates;
+                        dual_info("Per class constant mapping files is enabled");
```
Should be "are enabled".

jfdenise replied:
Thanks, done.

tbitonti commented on org.eclipse.transformer/src/main/java/org/eclipse/transformer/Transformer.java:
```diff
@@ -945,6 +950,33 @@ public boolean setRules() throws IOException, URISyntaxException, IllegalArgumen
+
+                    } else {
+                        perClassConstantStrings = null;
+                        dual_info("Per class constant mapping is not enabled");
```
Should be "Per class constant mapping files are not enabled"

jfdenise replied:
Thanks, done.

tbitonti commented on org.eclipse.transformer/src/main/java/org/eclipse/transformer/Transformer.java:
```diff
@@ -283,7 +283,11 @@ public static Options build(OptionSettings[] settings) {
-			OptionSettings.NO_GROUP);
+			OptionSettings.NO_GROUP),
+                
+                RULES_PER_CLASS_CONSTANT("tcd", "per-class-constant", "Transformation per class constant string replacements",
```
Remember to add this option to the Jakarta transformer and to the rules documentation.

jfdenise replied:
Thanks, done (added also to Maven plugin).

tbitonti commented on org.eclipse.transformer/src/main/java/org/eclipse/transformer/action/impl/ClassActionImpl.java:
```diff
@@ -950,6 +950,9 @@ private Object transformConstantValue(Object inputValue) {
 				if (outputString == null) {
 					transformCase = "direct";
 					outputString = transformDirectString(inputString);
+                                        if (outputString == null) {
```
Add a "transformCase" tag for this case, for example, "direct per class".

jfdenise replied:
Thanks, done.

tbitonti commented on org.eclipse.transformer/src/main/java/org/eclipse/transformer/action/impl/ClassActionImpl.java:
```diff
@@ -1132,7 +1135,16 @@ private Object transformElementValue(Object inputValue) {
 
 		} else {
+                    if (inputValue instanceof String) {
+                        String str = (String) inputValue;
```
I'd rename "str" to "inputString".

jfdenise replied:
Thanks, done.

tbitonti commented on org.eclipse.transformer/src/main/java/org/eclipse/transformer/action/impl/ClassActionImpl.java:
```diff
@@ -1250,6 +1262,9 @@ private int transform(MutableConstantPool constants) throws TransformException {
 								transformCase = "Direct";
 								outputUtf8 = transformDirectString(inputUtf8);
+                                                                if (outputUtf8 == null) {
+                                                                    outputUtf8 = transformConstantString(inputUtf8, inputName);
```
Add a "transformCase" assignment, for example, "Direct per class".

jfdenise replied:
Thanks, done.

tbitonti commented on org.eclipse.transformer/src/main/java/org/eclipse/transformer/action/impl/ClassActionImpl.java:
```diff
@@ -1278,6 +1293,9 @@ private int transform(MutableConstantPool constants) throws TransformException {
 							transformCase = "Direct";
 							outputString = transformDirectString(inputString);
+                                                        if (outputString == null) {
+                                                            outputString = transformConstantString(inputString, inputName);
```
Add a transformCase assignment, for example, "Direct per class".

jfdenise replied:
Thanks, done.

tbitonti commented on org.eclipse.transformer/src/main/java/org/eclipse/transformer/action/impl/SignatureRuleImpl.java:
```diff
@@ -181,6 +197,37 @@ public BundleData getBundleUpdate(String symbolicName) {
+                // Do we have a direct mapping?
+                String full = m.get(initialValue);
+                if (full != null) {
+                    System.out.println("DIRECT REPLACEMENT:[" + clazz + "]" + initialValue + "=>" + full);
```
The presumption is that this is a temporary output, and that the usual logger will eventually be used.

jfdenise replied:
Yes, Done.

tbitonti commented on org.eclipse.transformer/src/main/java/org/eclipse/transformer/action/impl/SignatureRuleImpl.java:
```diff
@@ -181,6 +197,37 @@ public BundleData getBundleUpdate(String symbolicName) {
+                String transformedString = initialValue;
+                boolean transformed = false;
+                // Transform all tokens possibly present in the inputValue
+                for (String k : m.keySet()) {
```
Should we keep looking after finding the first match?

tbitonti commented on org.eclipse.transformer/src/main/java/org/eclipse/transformer/action/impl/SignatureRuleImpl.java:
```diff
@@ -181,6 +197,37 @@ public BundleData getBundleUpdate(String symbolicName) {
+                boolean transformed = false;
+                // Transform all tokens possibly present in the inputValue
+                for (String k : m.keySet()) {
+                    if (transformedString.contains(k)) {
```
I'd add a comment here to indicate that the 'contains' test might be changed to use regular expressions in the future.

jfdenise replied:
Thanks, done.

tbitonti commented on org.eclipse.transformer/src/main/java/org/eclipse/transformer/action/impl/ClassActionImpl.java:
```diff
@@ -377,7 +377,7 @@ public ByteData apply(String inputName, byte[] inputBytes, int inputLength) thro
 		while (fields.hasNext()) {
 			FieldInfo inputField = fields.next();
-			FieldInfo outputField = transform(inputField, FieldInfo::new, SignatureType.FIELD);
+			FieldInfo outputField = transform(inputField, FieldInfo::new, SignatureType.FIELD, inputName);
```
All of these 'inputName' parameters are ugly, but that's not the fault of this update, but rather of the need to keep actions stateless.  There is room to refactor how the actions manage state, which is more than should be put into this pull request.

tbitonti commented on org.eclipse.transformer/src/main/java/org/eclipse/transformer/action/impl/SignatureRuleImpl.java:
```diff
@@ -181,6 +197,37 @@ public BundleData getBundleUpdate(String symbolicName) {
+        @Override
+        public String getConstantString(String initialValue, String clazz) {
+            Map<String, String> m = perClassConstantStrings.get(clazz);
+            if (m != null) {
```
This might immediately return (with null) if no class specific table ('m') is retrieved.  That would remove a level of indentation from the method, and would de-couple the "no substitution because no table was found" case from the "no substitution because no keys were found in the input string" case.  But, not all folks agree with the style, so I'm OK with this the way it is.

jfdenise replied:
I like it better, I made the change you proposed.

tbitonti commented:
Hi; reviewed the initial update.  Looks fine.  I have a number of comments, above, but all of the comments are for minor issues.

jfdenise commented:
@tbitonti , thank-you for your review I fixed them all. I have also added a unit test. 

I am wandering if the count of constant changes is correct. We are counting it both in UTF8 and in ConstantString, so that doubles the number. I checked with Direct String and it is the same.
I added a comment in the test assertion.

tbitonti commented:
@bjhargrave This looks good to me.  Can you take a quick look before merging?

bjhargrave commented on org.eclipse.transformer/src/main/java/org/eclipse/transformer/action/impl/ClassActionImpl.java:
```diff
@@ -1132,7 +1132,16 @@ private Object transformElementValue(Object inputValue) {
 			return outputElementValues;
 
 		} else {
+                    if (inputValue instanceof String) {
```
Could we not follow the normal `else if (...) {` flow like the previous blocks?

bjhargrave replied:
We should also move this above the `else if (inputValue instanceof Object[])` so that all the scalar types are done before the array.

jfdenise replied:
Ok, done.

bjhargrave commented on org.eclipse.transformer.cli/src/main/java/org/eclipse/transformer/jakarta/JakartaTransformer.java:
```diff
@@ -36,6 +36,7 @@ public static void main(String[] args) throws Exception {
 	public static final String	DEFAULT_BUNDLES_REFERENCE		= "jakarta-bundles.properties";
 	public static final String	DEFAULT_DIRECT_REFERENCE		= "jakarta-direct.properties";
 	public static final String	DEFAULT_MASTER_TEXT_REFERENCE	= "jakarta-text-master.properties";
+        public static final String	DEFAULT_PER_CLASS_CONSTANT_MASTER_REFERENCE	= "jakarta-per-class-constant-master.properties";
```
The formatting is wrong. The formatter uses tabs and not space. Not to get into a war, but we should be consistent so that future changes don't reformat the code.

This comment applies to all the files in this PR.

jfdenise replied:
Thank-you, I reformatted all the changes.

bjhargrave commented on org.eclipse.transformer/src/main/java/org/eclipse/transformer/action/impl/SignatureRuleImpl.java:
```diff
@@ -41,10 +41,18 @@
 		Map<String, String> renames, Map<String, String> versions, Map<String, BundleData> bundleUpdates,
 		Map<String, Map<String, String>> masterTextUpdates, Map<String, String> directStrings) {
+            this(logger, renames, versions, bundleUpdates, masterTextUpdates, directStrings, Collections.emptyMap());
+        }
```
Do we really need to overload this constructor? We are not at release 1.0 yet, so just change the original constructor and all places which call it.

jfdenise replied:
Constructor removed.

bjhargrave commented:
Thanks!
